oc new-app/new-build: handle the case when an imagestream matches but has no tags

csrwng opened the pull request:
We now display an error when we get a match for the imagestream but the imagestream has no tags.
If the imagestream has other tags, then we display those partial matches and tell you how to specify them. 
You can still use an imagestream with no tags by specifying --allow-missing-imagestream-tags

Fixes #7613

csrwng commented:
@bparees ptal

bparees commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -684,6 +685,20 @@ func transformError(err error, c *cobra.Command, fullName string, groups errorGr
 		return
+	case newapp.ErrNoTagsFound:
+		buf := &bytes.Buffer{}
+		fmt.Fprintf(buf, "  Use --allow-missing-imagestream-tags to use this image stream\n\n")
```
to use this image stream tag

csrwng replied:
I did have that, but it didn't sound right if you weren't specifying a tag

csrwng replied:
and if you have no tags on the image stream, then you're really saying you want to use the image stream anyway.

bparees replied:
hm. ok i guess i'm convinced.

bparees commented on pkg/generate/app/errors.go:
```diff
@@ -39,7 +39,7 @@ func (e ErrPartialMatch) Error() string {
 }
 
-// UsageError is the usage error message returned when only a partial match is
+// Suggestion is the usage error message returned when only a partial match is
```
@rhcarvalho @PI-Victor this is a good example of fixing minor issues w/o a separate PR/commit.

bparees commented on pkg/generate/app/imagestreamlookup.go:
```diff
@@ -90,17 +113,25 @@ func (r ImageStreamSearcher) Search(precise bool, terms ...string) (ComponentMat
-						Meta:        meta,
-					})
+					if r.AllowMissingTags {
+						addMatch(finalTag, score, nil, false)
```
hm.  we should only be doing this if the imagestream score was 0.0, i think?  (wondering what happens if you partially match an imagestream that has no tags, or at least doesn't have this tag)

csrwng commented on pkg/generate/app/imagestreamlookup.go:
```diff
@@ -75,6 +76,28 @@ func (r ImageStreamSearcher) Search(precise bool, terms ...string) (ComponentMat
+						Value:       term,
+						Argument:    fmt.Sprintf("--image-stream=%q", name),
+						Name:        name,
+						Description: fmt.Sprintf("Image stream %q (tag %q) in project %q", stream.Name, tag, stream.Namespace),
```
this is wrong

csrwng commented:
@bparees I added another commit to better handle partial results. Given the following image streams:

```
$ oc get is
NAME      DOCKER REPO                            TAGS            UPDATED
ruby1     172.30.173.105:5000/myproject/ruby1    tag2,tag1,tip   21 minutes ago
ruby2     172.30.173.105:5000/myproject/ruby2
ruby31    172.30.173.105:5000/myproject/ruby31
ruby32    172.30.173.105:5000/myproject/ruby32
```

Here's the output for the cases we discussed:
1) Partial matches / no tags in each matching stream:
```
$ oc new-app --dry-run ruby3
error: multiple images or templates matched "ruby3": 2

The argument "ruby3" could apply to the following Docker images or OpenShift image streams:

* Image stream "ruby31" in project "myproject"
  Use --image-stream="myproject/ruby31" --allow-missing-imagestream-tags to specify this image or template

* Image stream "ruby32" in project "myproject"
  Use --image-stream="myproject/ruby32" --allow-missing-imagestream-tags to specify this image or template
```

2) Partial matches, some tags do exist
```
$ oc new-app --dry-run rub
error: multiple images or templates matched "rub": 7

The argument "rub" could apply to the following Docker images or OpenShift image streams:

* Image stream "ruby1" (tag "tag1") in project "myproject"
  Use --image-stream="myproject/ruby1:tag1" to specify this image or template

* Image stream "ruby1" (tag "tag2") in project "myproject"
  Use --image-stream="myproject/ruby1:tag2" to specify this image or template

* Image stream "ruby1" (tag "tip") in project "myproject"
  Use --image-stream="myproject/ruby1:tip" to specify this image or template

* Image stream "ruby" (tag "2.2") in project "openshift"
  Use --image-stream="openshift/ruby:2.2" to specify this image or template

* Image stream "ruby2" in project "myproject"
  Use --image-stream="myproject/ruby2" --allow-missing-imagestream-tags to specify this image or template

* Image stream "ruby31" in project "myproject"
  Use --image-stream="myproject/ruby31" --allow-missing-imagestream-tags to specify this image or template

* Image stream "ruby32" in project "myproject"
  Use --image-stream="myproject/ruby32" --allow-missing-imagestream-tags to specify this image or template
```

3) Exact match, no tags exist
```
$ oc new-app --dry-run ruby2
error: no tags found on matching image stream: "myproject/ruby2"

The image stream "ruby2" exists, but it has no tags.

  Use --allow-missing-imagestream-tags to use this image stream
```

4) Exact match, tag exists
```
$ oc new-app --dry-run ruby1:tip
--> Found image f60c63f (2 days old) in image stream ruby1 under tag "tip" for "ruby1:tip"

    Ruby 2.2
    --------
    Platform for building and running Ruby 2.2 applications

    Tags: builder, ruby, ruby22

    * This image will be deployed in deployment config "ruby1"
    * Port 8080/tcp will be load balanced by service "ruby1"
      * Other containers can access this service through the hostname "ruby1"

--> Creating resources with label app=ruby1 ...
--> Success (DRY RUN)
```

The reason we were not getting matches from all namespaces when only matching partially was because the matches were getting scored differently. So searching for "rub" (no tag), internally we choose "latest" as the tag. So the imagestream in openshift has a lower score (better match) than those in the local namespace, because the tag matches "latest". If there's multiple matches but one has a lower score than the rest, the resolver will return a partial match error only with that result. What we weren't taking into consideration is that you didn't specify that "latest" tag, we picked it for you. I've fixed it now so that if you didn't specify the tag, then matches on other tags get the same score.

bparees commented:
that output looks good to me, what happens in this variant of scenario 3:

`oc new-app --dry-run ruby2:foo`

?

i'd assume it's the same output, but just want to confirm.

csrwng commented:
hmm, we include the tag in the output, which we shouldn't... 

```
$ oc new-app --dry-run ruby2:foo
error: no tags found on matching image stream: "myproject/ruby2"

The image stream "ruby2:foo" exists, but it has no tags.

  Use --allow-missing-imagestream-tags to use this image stream
```

I'll fix it.

csrwng commented:
ok, better now:
```
$ oc new-app --dry-run ruby2:foo
error: no tags found on matching image stream: "myproject/ruby2"

The image stream "myproject/ruby2" exists, but it has no tags.

  Use --allow-missing-imagestream-tags to use this image stream
```

bparees commented:
and what about an imagestream that flat out doesn't exist?

`oc new-app --image-stream=foobar` ?

bparees commented:
lgtm, squash please.

csrwng commented:
squashed [merge]
